File: agent/native-cpp/src/tools/tool_registry.cpp
#include "agent/tools/tool_registry.h"

#include <utility>

namespace atrium::agent {

bool ToolRegistry::registerTool(ToolSpec spec, ToolHandler handler) {
    if (spec.name.empty() || !handler) {
        return false;
    }

    const auto name = spec.name;
    auto [it, inserted] = m_tools.emplace(name, Entry{std::move(spec), std::move(handler)});
    (void)it;
    return inserted;
}

bool ToolRegistry::contains(const std::string &name) const {
    return m_tools.find(name) != m_tools.end();
}

ToolResult ToolRegistry::call(const ToolCall &call) const {
    auto it = m_tools.find(call.name);
    if (it == m_tools.end()) {
        return ToolResult{false, "", "tool not registered: " + call.name};
    }
    return it->second.handler(call);
}

std::vector<ToolSpec> ToolRegistry::list() const {
    std::vector<ToolSpec> specs;
    specs.reserve(m_tools.size());
    for (const auto &[name, entry] : m_tools) {
        (void)name;
        specs.emplace_back(entry.spec);
    }
    return specs;
}

} // namespace atrium::agent

File: agent/src/tools/toolRegistry.ts
export interface ToolSpec {
  name: string;
  description: string;
  inputSchemaJson: string;
}

export interface ToolCall {
  name: string;
  argumentsJson: string;
}

export interface ToolResult {
  ok: boolean;
  content: string;
  error: string;
}

export type ToolHandler = (call: ToolCall) => ToolResult | Promise<ToolResult>;

interface ToolEntry {
  spec: ToolSpec;
  handler: ToolHandler;
}

export class ToolRegistry {
  readonly #tools = new Map<string, ToolEntry>();

  registerTool(spec: ToolSpec, handler: ToolHandler): boolean {
    if (spec.name.length === 0) {
      return false;
    }
    if (this.#tools.has(spec.name)) {
      return false;
    }

    this.#tools.set(spec.name, { spec: structuredClone(spec), handler });
    return true;
  }

  contains(name: string): boolean {
    return this.#tools.has(name);
  }

  async call(call: ToolCall): Promise<ToolResult> {
    const entry = this.#tools.get(call.name);
    if (!entry) {
      return { ok: false, content: "", error: `tool not registered: ${call.name}` };
    }
    return entry.handler(call);
  }

  list(): ToolSpec[] {
    return Array.from(this.#tools.values(), (entry) => structuredClone(entry.spec));
  }
}

